src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.Cache.cs

```diff
@@ -566,37 +568,25 @@ internal void UpdateSortedParameterCache(ref ReadStackFrame frame)
             frame.CtorArgumentState.ParameterRefCache = null;
         }
 
-        internal void InitializeSerializePropCache()
+        internal void InitializePropCache()
         {
             Debug.Assert(PropInitFunc != null);
             Debug.Assert(Options._context != null);
 
-            PropertyCacheArray = PropInitFunc(Options._context);
-        }
-
-        internal void InitializeDeserializePropCache()
-        {
-            if (PropertyCacheArray == null)
-            {
-                InitializeSerializePropCache();
-            }
-
-            Dictionary<string, JsonPropertyInfo> propertyCache = new(Options.PropertyNameCaseInsensitive
-                ? StringComparer.OrdinalIgnoreCase
-                : StringComparer.Ordinal);
+            JsonPropertyInfo[] array = PropInitFunc(Options._context);
 
-            for (int i = 0; i < PropertyCacheArray!.Length; i++)
+            var properties = new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, array.Length);
+            for (int i = 0; i < array.Length; i++)
             {
-                JsonPropertyInfo jsonPropertyInfo = PropertyCacheArray[i];
-
-                if (!JsonHelpers.TryAdd(propertyCache, jsonPropertyInfo.NameAsString, jsonPropertyInfo))
+                JsonPropertyInfo property = array[i];
+                if (!properties.TryAdd(property.NameAsString, property))
```

Comment:
How do we detect duplicate property names when we have less than 10 (dictionary-creation threshold) properties on a type, since we don't create a dictionary in those cases (per `JsonPropertyDictionary` implementation)?

[I know we do detect this based on [tests](https://github.com/dotnet/runtime/blob/8bcc982b489f4058dbdb6214b7536ce7c7d83ba2/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PropertyNameTests.cs#L125)]

Reply:
The Add() and Set() methods loop through the list and manually look for duplicates via `ContainsProperty()`